sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -797,6 +834,19 @@ object NullPropagation extends Rule[LogicalPlan] {
       // a null literal.
       case e: NullIntolerant if e.children.exists(isNullLiteral) =>
         Literal.create(null, e.dataType)
+
+      // [SPARK-36665] Unwrap inside of IsNull/IsNotNull if the inside is NullIntolerant
+      // E.g. IsNull(Not(null)) == IsNull(null)
+      // Cannot apply to `ExtractValue` as the query planner uses the trait to resolve the columns.
+      // E.g. the planner may resolve column `a` to `a#123`, then IsNull(a#123) cannot be optimized
+      // UnaryExpression only for now as applying this optimization to other expressions is too
+      // disruptive for some tests (e.g. [SPARK-32290].) TODO remove e.isInstanceOf[UnaryExpression]
+      case IsNull(e: NullIntolerant) if e.isInstanceOf[UnaryExpression] &&
+        !e.isInstanceOf[ExtractValue] && e.children.nonEmpty =>
+        e.children.map(IsNull(_): Expression).reduceLeft(Or)
```

Comment:
can we simplify this? shouldn't `UnaryExpression` only have one child?

Reply:
I was thinking about making it to future proof so that we can remove the `UnaryExpression` restriction.

Reply:
I see. Make sense.